### ISSUE TITLE
Enhancement: Add serialization info to parser

### ISSUE THREAD

**Issue by nzakas** [typescript-eslint/typescript-eslint#6541]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Up until now, the flat config in ESLint was unable to serialize custom parsers, meaning that caching and `--print-config` would not work with `typescript-eslint/parser`. I just [added a feature](https://github.com/eslint/eslint/pull/16944) that allows parsers to export a `meta` key that would provide serialization information, such as:

```js
exports.meta = {
    name: "typescript-eslint/parser",
    version: "0.0.0"
};
```

This should be exported on the same object as `parse()` and `parseForESLint()`.

### Fail

```typescript
// not applicable
```

### Pass

```typescript
// not applicable
```

### Additional Info

It seems like there may be some fields meant for rules in this template. :)

**Comment by brettz9** [comment]
FWIW, this would be useful for us in `eslint-plugin-jsdoc` to auto-detect from the parser metadata that the user wants TypeScript mode as far as JSDoc's.

**Comment by bradzacher** [comment]
@brettz9 - you can actually get this (in a way) in lint rules today via `context.parserPath`.
The metadata would help to concretely identify it (eg you could do `require(context.parserPath).meta.name`), but should still be able to make an educated guess by just inspecting the path.

**Comment by brettz9** [comment]
@bradzacher : Yes, thank you, that's what we have been using, but for a user trying the new flat config, it seems it is not working per https://github.com/eslint/eslint/issues/16878#issuecomment-1425968326 .

**Comment by nzakas** [comment]
You should also be able to do `context.parser.meta.name` with flat config.
